keps/sig-network/2004-topology-aware-subsetting/README.md

```diff
@@ -0,0 +1,580 @@
+costs.
+  - Performance: Traffic within a zone usually has less latency and bandwidth
+constraints, having a better performance than traffic leaving the zone.
+
+In this KEP we are going to focus on avoiding cross-zone traffic when in-zone
+endpoints would suffice. We're attempting to provide a simple and more automatic
+approach to topology aware routing. This API will still allow users to indicate
+that they prefer to keep traffic in the same zone if there's sufficient
+capacity. With this approach users won't have to configure anything by default
+for most use cases.
+
+
+### Goals
+- Provide a simple way for users to indicate their preference for keeping
+  traffic in zone.
+- Use the standardized topology label `topology.kubernetes.io/zone` to derive
+  the zones of nodes and endpoints.
+- Use EndpointSlice subsetting as the primary mechanism for topology aware
+  routing.
+  - Improve kube-proxy scalability by limiting the number of endpoints it needs
+    to process.
+- Minimize churn of EndpointSlices while doing topology aware distribution.
+- Minimize the number of new EndpointSlices required.
+- Provide a simple API that requires minimal configuration for most users.
+
+### Non-Goals
+- Real-time distribution rebalancing based on traffic load or distribution
+  feedback or metrics.
+- Providing a way for kube-proxy to consume subsets of EndpointSlices. (This is
+  covered by [KEP 2030](https://github.com/kubernetes/enhancements/issues/2030)).
+- Providing a way to require Node local traffic ([this is a separate
+  KEP](https://github.com/kubernetes/enhancements/pull/1944)).
+- Multi-cluster topology aware routing (this same pattern may be useful there
+  though).
+- Region based topology aware routing (this may come later).
+- Ensuring that Pods are distributed evenly across zones.
+
+## Proposal
+
+### Terminology
+
+- **EndpointSlice Group**: One or multiple EndpointSlices that will be consumed
+  by the same set of zones.
+- **Endpoints**: In this KEP, “endpoints” are used to describe individual Pod
+  IPs. These are not references to Kubernetes Endpoints resources.
+
+### Configuration
+
+This proposal builds off of the [EndpointSlice API](https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/).
+
+We propose a new Service annotation that would act as a hint for the
+EndpointSlice controller: `endpointslice.kubernetes.io/subsetting`. If set to
+`Auto` the EndpointSlice controller will conduct a more conservative routing
+mechanism. It will keep traffic in zone as long as each zone has a sufficient
+number of endpoints. If a zone does not have a sufficient number of endpoints,
+traffic will be routed to other zones. This will be explained in more detail
+below.
+
+If this annotation is unspecified or empty, the EndpointSlice controller will
+continue to behave as it already does with no form of subsetting.
+
+#### Interoperability
+
+This annotation may not be set on a Service if the `topologyKeys` field is set.
```

Comment:
Do we have any precedent for failing to run if two incompatible feature gates are specified?

Reply:
I don't think so, I've made this language a bit more realistic.